docs/src/test/java/org/elasticsearch/smoketest/DocsClientYamlTestSuiteIT.java

```diff
@@ -97,6 +99,23 @@ protected ClientYamlTestClient initClientYamlTestClient(
         return new ClientYamlDocsTestClient(restSpec, restClient, hosts, esVersion, masterVersion, this::getClientBuilderWithSniffedHosts);
     }
 
+    @After
+    public void cleanup() throws Exception {
+        if (isMachineLearningTest() || isDataFrameTest()) {
+            ESRestTestCase.waitForPendingTasks(adminClient());
```

Comment:
I wonder how bad it'd be to do this after *every* test. I don't feel great about relying on stuff in the test name. It just feels a bit too magical.

Reply:
It is a little bit complicated because Rollups do the wait in the base [ESRestTestCase](https://github.com/elastic/elasticsearch/blob/master/test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java#L468)

Additionally some tests leave tasks running. `get-follow-info.asciidoc` line 38 is a good example as it creates various CCR tasks which will be waited on indefinitely unless the test teardown is run. Interestingly what appears to be happening is the `@After` method of this class is called before the test teardown

Reply:
>  Interestingly what appears to be happening is the `@After` method of this class is called before the test teardown

Weird!

I'm not a big fan of leaving things running in those tests either. Is there a way you could do something like the rollups here? It looks like it only cares about rollup style jobs. Does ml have something similar?

Reply:
Yeah rollups filter the waiting tasks with `taskName.startsWith("xpack/rollup/job")` and we can do something similar with ml jobs but the action causing the leakage in #43271 is indexing a document not an ml task. Waiting for all tasks catches unexpected issues and actually helps debugging tests that have failed due to leakage from a previous test, experience from using this in `XPackRestIT` has shown that it is very valuable.    

If I remove the `if (isMachineLearningTest() || isDataFrameTest()) {` check then the tests that fail with pending tasks are ccr and rollup. I'll look into what's happening there and maybe there is a way of removing the _if ml ... _ conditional

Reply:
I took a look at the Rollup and CCR tests, unfortunately it is not possible to wait for pending tasks after every test because those tests require special handling. I cannot see a way to simplify the logic and I think the current code is best as it is explicitly for the ml & data frame tests.

Also as more xpack feature snippet testing is added I would expect more usages of the pattern e.g. `if (isSecurityTest()) { // security specific cleanup` 
 
Using the test name to determine if the test is an ml test is a valid use. [XPackRestIT](https://github.com/elastic/elasticsearch/blob/ccb5d934363c37506b76119ac050a254fa80b5e7/x-pack/plugin/src/test/java/org/elasticsearch/xpack/test/rest/XPackRestIT.java#L324) set the precedent some time ago and it has not caused problems there.

Reply:
I'm *really* not a fan of looking at the test name. I know XPackRestIT does it and I think it is sneaky black magic that will cause tests to fail in very difficult ways to trace. One badly named test invoking ml will cause subsequent tests to fail. Sometimes. Randomly.

I'm ok with merging this, but I'd really like a follow up issue to remove it *somehow*. Because I'm 100% sure somebody is going to lose many hours to debugging errors caused by a funny named test one day.

Reply:
Can you detect a data frame test or ML test by *looking* at the public API somehow? Like by looking for jobs or something.....